pynest/nest/tests/test_connect_arrays_mpi.py

```diff
@@ -0,0 +1,142 @@
+
+
+@unittest.skipIf(not HAVE_MPI, 'NEST was compiled without MPI')
+@unittest.skipIf(not HAVE_MPI4PY, 'mpi4py is not available')
+class ConnectArraysMPICase(unittest.TestCase):
+    """
+    This TestCase uses mpi4py to collect and assert results from all
+    processes, and is supposed to only be run with multiple processes.
+    If running with nosetests or pytest, this TestCase is ignored when
+    run with a single process, and called from TestConnectArraysMPI using
+    multiple processes.
+    """
+    non_unique = np.array([1, 1, 3, 5, 4, 5, 9, 7, 2, 8], dtype=np.uint64)
+    comm = MPI.COMM_WORLD
+
+    # With nosetests, only run these tests if using multiple processes
+    __test__ = MULTIPLE_PROCESSES
+
+    def assert_connections(self, expected_sources, expected_targets, rule='one_to_one'):
+        """Gather connections from all processes and assert against expected connections"""
+        conns = nest.GetConnections()
+        projections = [[s, t] for s, t in zip(conns.source, conns.target)]
+        if rule == 'one_to_one':
+            expected_projections = np.array([[s, t] for s, t in zip(expected_sources, expected_targets)])
+        elif rule == 'all_to_all':
+            expected_projections = np.array([[s, t] for s in expected_sources for t in expected_targets])
+        else:
+            self.assertFalse(True, 'rule={} is not valid'.format(rule))
+
+        recv_projections = self.comm.gather(projections, root=0)
+        if self.comm.Get_rank() == 0:
+            # Flatten the projection lists to a single list of projections
+            recv_projections = np.array([proj for part in recv_projections for proj in part])
+            # Results must be sorted to make comparison possible
+            np.testing.assert_array_equal(np.sort(recv_projections, axis=0), np.sort(expected_projections, axis=0))
+        else:
+            self.assertIsNone(recv_projections)
+
+    def setUp(self):
+        nest.ResetKernel()
+
+    def test_connect_arrays_unique(self):
+        """Connecting NumPy arrays of unique node IDs with MPI"""
+        n = 10
+        nest.Create('iaf_psc_alpha', n)
+        sources = np.arange(1, n+1, dtype=np.uint64)
+        targets = np.arange(1, n+1, dtype=np.uint64)
+        weights = 1.5
+        delays = 1.4
+
+        nest.Connect(sources, targets, syn_spec={'weight': weights, 'delay': delays})
+
+        self.assert_connections(sources, targets, rule='all_to_all')
+
+    def test_connect_arrays_nonunique(self):
+        """Connecting NumPy arrays with non-unique node IDs with MPI"""
+        n = 10
+        nest.Create('iaf_psc_alpha', n)
+        sources = np.arange(1, n+1, dtype=np.uint64)
+        targets = self.non_unique
+        weights = np.ones(n)
+        delays = np.ones(n)
+        nest.Connect(sources, targets, syn_spec={'weight': weights, 'delay': delays},
+                     conn_spec='one_to_one')
+
+        self.assert_connections(sources, targets)
```

Comment:
the fact that the default value for "rule" in ``Connect`` is "all_to_all" but it is "one_to_one" for ``assert_connections`` seems like pure evil x)
I would make a variable for the rule every time and always specify it explicitly (also removing the default value in the function argument)

Reply:
Done!